Initial Kubernetes guide for Google Cloud. (#1317)
micronaut-projects/micronaut-guides: guides/micronaut-k8s-gcp/micronaut-k8s-gcp.adoc

common:header-top.adoc[]

== Getting Started

In this guide, we will deploy three microservices on the https://cloud.google.com/kubernetes-engine[Google Kubernetes Engine (GKE)]. We will use Kubernetes Service discovery and Distributed configuration to wire up our microservices.

You will discover how the Micronaut framework eases Kubernetes integration and deployment to GKE.

common:costs-gcloud.adoc[]

external:micronaut-k8s/requirements.adoc[]
common:jq[]

> * You need a https://cloud.google.com/gcp/[Google Cloud Platform (GCP)] account and a GCP project.

Reviewer: Does this require billing to be enabled?

external:micronaut-k8s/create-app-external-guides.adoc[]

common:gcp.adoc[]

common:gcp-billing.adoc[]

common:gcp-docker.adoc[]

=== Create GKE cluster

> In this guide we will create cluster with name `micronaut-k8s` and we will use 2x `n1-standard-2` instances for GKE nodes. Everything will be deployed on your default Google Cloud region. You can check your default region by running:

Reviewer: ```suggestion
In this guide we will create a cluster named `micronaut-k8s` and we will use two `n1-standard-2` instances as GKE nodes. Everything will be deployed on your default Google Cloud region. You can check your default region by running:
```

[source, bash]
----
> gcloud config get-value compute/region
----

Create `micronaut-k8s` kubernetes cluster:

[source, bash]
----
gcloud container clusters create micronaut-k8s --machine-type n1-standard-2 --num-nodes 2
----

== Prepare and Deploy Microservices

> Define `GCP_PROJECT_ID` environment variable to store your GCP project id.

Reviewer: ```suggestion
Set the `GCP_PROJECT_ID` environment variable to store your GCP project id.
```

[source,bash]
----
export GCP_PROJECT_ID="$(gcloud config get-value project)"
----

=== Users Microservice

external-template:micronaut-k8s-gcp/export-repository.adoc[arg0=users]
external-template:micronaut-k8s-gcp/k8s-microservice.adoc[arg0=users]
external-template:micronaut-k8s-gcp/docker-microservice.adoc[arg0=users]

=== Orders Microservice

external-template:micronaut-k8s-gcp/export-repository.adoc[arg0=orders]
external-template:micronaut-k8s-gcp/k8s-microservice.adoc[arg0=orders]
external-template:micronaut-k8s-gcp/docker-microservice.adoc[arg0=orders]

=== API Microservice

external-template:micronaut-k8s-gcp/export-repository.adoc[arg0=api]
external-template:micronaut-k8s-gcp/k8s-microservice.adoc[arg0=api]
external-template:micronaut-k8s-gcp/docker-microservice.adoc[arg0=api]

=== Deploy Services to GKE

Create a directory for `kubectl` configuration.

[source,bash]
----
mkdir -p $HOME/.kube
----

Install kubectl component.

[source, bash]
----
gcloud components install kubectl
----

Generate `kubectl` configuration for authentication to GKE.

[source, bash]
----
gcloud container clusters get-credentials micronaut-k8s
----

> Set `KUBECONFIG` to the created config file. This variable is consumed by `kubectl`.

Reviewer: ```suggestion
Set the value of `KUBECONFIG` to the created config file. This variable is consumed by `kubectl`.
```

[source,bash]
----
export KUBECONFIG=$HOME/.kube/config
----

Deploy the _auth.yml_ file that we created in the guideLink:micronaut-k8s[Kubernetes and the Micronaut Framework] guide.

[source,bash]
----
kubectl apply -f auth.yml
----

> Create `gcpsersecret` secret that will be used for authentication to Google Cloud Container Registry. This is needed because GKE needs credentials to be able to pull microservices images.

Reviewer: ```suggestion
Create a secret named `gcpsersecret` that will be used for authentication to Google Cloud Container Registry. This is needed because GKE needs credentials to be able to pull microservices images.
```

[source,bash]
----
kubectl get pods -n=micronaut-k8s
----

[source,text]
----
NAME                      READY   STATUS    RESTARTS   AGE
api-6b884d9c88-hxsdk      1/1     Running   0          21s
orders-54465f845c-k7z5t   1/1     Running   0          24s
users-d8b46cf48-wq6sm     1/1     Running   0          29s
----

Run the next command to check the status of the microservices:

[source,bash]
----
kubectl get services -n=micronaut-k8s
----

[source,text]
----
NAME     TYPE           CLUSTER-IP     EXTERNAL-IP      PORT(S)          AGE
api      LoadBalancer   10.112.14.49   <redacted>       8080:31299/TCP   37s
orders   NodePort       10.112.2.71    <none>           8080:30613/TCP   40s
users    NodePort       10.112.2.227   <none>           8080:32718/TCP   44s
----

NOTE: `If EXTERNAL-IP` is in <pending> state wait a couple of seconds and then run command again.

Run the next command to retrieve the URL of the `api` microservice:

[source,bash]
----
export API_URL=http://$(kubectl get svc api -n=micronaut-k8s -o json | jq -r '.status.loadBalancer.ingress[0].ip'):8080
----

external:micronaut-k8s/verify.adoc[]

== Cleaning Up

> To delete all resources that were created in this guide run next command.

Reviewer: ```suggestion
To delete all resources that were created in this guide run next the command.
```

[source,bash]
----
kubectl delete namespaces micronaut-k8s
----

external-template:micronaut-k8s-gcp/gcp-delete-repository.adoc[arg0=users]
external-template:micronaut-k8s-gcp/gcp-delete-repository.adoc[arg0=orders]
external-template:micronaut-k8s-gcp/gcp-delete-repository.adoc[arg0=api]

> Run the next command to delete `micronaut-guide` GKE cluster:

Reviewer: ```suggestion
Run the next command to delete the `micronaut-guide` GKE cluster:
```

== Cleaning Up

> To delete all resources that were created in this guide run next the command.

Reviewer: ```suggestion
To delete all resources that were created in this guide run the next command.
```
